### ISSUE TITLE
Downward compatibility pgBackRest repository host to DB servers

### ISSUE THREAD

**Issue by Skado003** [pgbackrest/pgbackrest#2294]
Please provide the following information when submitting an issue (feature requests or general comments can skip this):

1. pgBackRest version: 2.49

2. PostgreSQL version: 15.5

3. Operating system/version - if you have more than one server (for example, a database server, a repository host server, one or more standbys), please specify each: both DB and repository host: Ubuntu 22.04.3 LTS

4. Did you install pgBackRest from source or from a package? Package

5. Please attach the following as applicable:
    - `pgbackrest.conf` file(s)
    - `postgresql.conf` settings applicable to pgBackRest (`archive_command`, `archive_mode`, `listen_addresses`, `max_wal_senders`, `wal_level`, `port`)
    - errors in the postgresql log file before or during the time you experienced the issue
    - log file in `/var/log/pgbackrest` for the commands run (e.g. `/var/log/pgbackrest/mystanza_backup.log`)

7. Describe the issue: 
Question: 
For the repository host it is mentioned in documentation: 
"NOTE:
The pgBackRest version installed on the repository host must exactly match the version installed on the PostgreSQL host." 
We make use of multiple DB servers and 1 repository host. DB servers are patched for PostgreSQL and pgBackRest packages at different times. This leads to issues with backup. It would be very helpfull when there would be some kind of downwards compatibility. So where the pgbackrest version on repository is allowed to be (slightly) higher then the ones on the DB servers.
Is this something which can be implemented?

**Comment by dwsteele** [comment]
> Is this something which can be implemented?

In theory, yes, but only at a major cost to feature work. As implemented, the remotes are very "smart" which means that wide variety of changes are protocol breaking changes. So, we'd need to make the remotes less functional or somehow version all those changes. Last time I surveyed this we were making breaking changes at least every other release or so.

We've been working to make the remote functionality more isolated, but there are still some big protocol changes coming in this area, all of which would be virtually impossible to make backward compatible.

So, in short, this is something that we are planning/hoping for but I don't see it happening any time soon.

**Comment by ankel** [comment]
@dwsteele  would you consider supporting semantic versioning for backrest, ie. all 2.x versions are api-compatible with each other, and api-breaking changes would only happen between 2.x and 3.y?

**Comment by dwsteele** [comment]
> all 2.x versions are api-compatible with each other, and api-breaking changes would only happen between 2.x and 3.y

Again, this can only happen with a lot more development resources than we have. Right now most major features require some protocol changes so we'd only be able to put bug fixes in the older version, i.e. 2.x would die pretty much as soon as 3.x is started. I'd prefer to just number our releases 100, 101, 102, etc., but I haven't managed to convince all the devs that is a good idea.

If you need the protocol to be stable then pin a version you like and stick with it. If you want new features then updating pgbackrest on all hosts is just part of the drill.
